### DISCUSSION TITLE
letsencrypt certificate expired - how do i generate and configure a new for modoboa

### DISCUSSION THREAD

**Discussion by CyberfoxDK** [modoboa/modoboa#2711] [Upgrade]
Very simple question, but not so simple a solution.

My certificate expired...   how do i create and assign a new certificate (letsencrypt or other) to modoboa ?

**Comment by tonioo** [comment]
Your certificate should be renewed automatically, there is a dedicated cron job for that

**Reply by CyberfoxDK** [reply]
First - I did not know that.
Second - i have blocked port 80 to the webserver, and because of this i suspect that the automatically renewal did not go through.
Is there a way to trigger a new letsencrypt certificate on an existing system, or would it be faster to just migrate the mailboxes to a new installation ?

(Thanks for a great product Tonioo !)

**Reply by nicosqc** [reply]
You can do ``certbot renew`` to automatically renew existing certificates. If it doesn't work, try adding your webserver after a double dash like ``certbot renew --apache`` or ``certbot renew --nginx``. And yes, Port 80 needs to be open for this if you use the default method over http; there is a more advanced option to authenticate using TXT-records in your DNS but this is a lot more effort to automate.

**Reply by CyberfoxDK** [reply]
But i assume that renewal would only work if the existing certificate is not yet expired - am i wrong on this ?

**Reply by nicosqc** [reply]
Yeah, it will work regardless. If its expired it just gets you a new certificate

**Reply by CyberfoxDK** [reply]
Thanks nicosqc - I did not know that either... will give it a try and get back with the results.

**Reply by nicosqc** [reply]
No worries, happy to help. Please note that I did not use the installer script to install the certificate and I do not know if the certbot binary is on PATH when installed by the installer script; if you get a command not found error you might have to put in the full path or load a python virtual environment before running certbot. You can check the crontab that tonioo mentioned to see how it is executed in your case, check the files in be in ``/etc/cron.d/`` or similiar directories.

**Reply by CyberfoxDK** [reply]
Thank you a bunch nicosqc !
While the commands initially gave an error, installing the nginx plugin (  sudo apt-get install python3-certbot-nginx  ) fixed the issue, and i was able to renew the expired certificate afterwords.
Appreciate it !

**Reply by nicosqc** [reply]
Glad this solved your issue. I'd still recommend to check how the renew is done in your cron files; it might have been installed using the python ``pip`` utility and you might have two versions of the certbot binary installed next to each other

**Reply by CyberfoxDK** [reply]
I will take a closer look at it, but think the issue was with verification of the token due to blocking of port 80.

**Reply by nicosqc** [reply]
This was certainly an issue when the command executing from crontab attempted to renew your certificate, I just mentioned it to make sure that the current configuration will be future proof. When the same package is installed from two different sources you might end up having issues upgrading either one of them, but I am just speculating here. If it works, it works (at least for now).

**Reply by CyberfoxDK** [reply]
And i have made some changes, allowing letsencrypt cidr to access the server on port 80. (hoping that they do not change iprange).
But yes, it seems to work, and just knowing that an expired certificate can be renewed helped a lot.
I thought that it was over, and that i had to request a completely new certificate with new public and private keys...  so this helps me a lot.

**Reply by nicosqc** [reply]
**If you know what you're doing** and want to lock down your port 80 at all times unless when certbot is making a renew, you can think about commenting out/removing the ``certbot`` crontab and create a systemd service with a Pre- and Post-Execute hook that opens and closes Port 80. **I recommend against this approach as explained further below and in the following reply.** Certbot itself has a similar functionality of pre- and post-renew hooks [(check the certbot docs here if you like)](https://eff-certbot.readthedocs.io/en/stable/using.html#pre-and-post-validation-hooks) which I havent messed with myself since creating a systemd service and timer was easier for me.

``/etc/systemd/system/certbot.timer``
```
[Unit]
Description=Run certbot twice daily

[Timer]
OnCalendar=*-*-* 00,12:00:00
RandomizedDelaySec=43200
Persistent=true

[Install]
WantedBy=timers.target
```
``/etc/systemd/system/certbot.service``
```
[Unit]
Description=Certbot
Documentation=file:///usr/share/doc/python-certbot-doc/html/index.html
Documentation=https://letsencrypt.readthedocs.io/en/latest/
[Service]
Type=oneshot
ExecStartPre=(open your Port 80 here with a script like ufw or iptables)
ExecStart=/usr/bin/certbot -q renew --nginx
ExecStartPost=/bin/systemctl reload nginx ; /bin/systemctl reload postfix ; /bin/systemctl reload dovecot ; (close your Port 80 here with a script like ufw or iptables)
PrivateTmp=true
```

I recommend to just redirect all traffic to HTTPS (as done by default in modoboa installer script). Since I run the same webserver on Port 443 I don't see how disallowing traffic on Port 80 is a significant gain in security. It is probably more of an inconvenience to your users since if they forget to explicity add ``https://`` in front they will get a ``connection refused`` error page, making them think something is wrong with the server. This is the vhost-configuration I have running:

```
server {
    listen 80;
    listen [::]:80;
    server_name mail.myserver.tld;
    rewrite ^ https://$server_name$request_uri? permanent;
}
```
Furthermore, you can choose to serve only 404 errors when accessing any page over http thats not explicitly covered as a virtual host (like access by IP):
``/etc/nginx/sites-available/default``
```
server {
    listen 80 default_server;
    listen [::]:80 default_server;

    server_name _;

    location / {
        return 404;
    }
}
```

**Reply by nicosqc** [reply]
I just checked my nginx files and if you restrict access to port 80, autoconfig will not work:
``/etc/nginx/sites-enabled/autoconfig.myserver.tld.conf``
```
# This file was automatically installed on 2022-06-20T12:08:51.089851
upstream automx {
    server unix:/run/uwsgi/app/automx_instance/socket fail_timeout=0;
}

server {
    listen 80;
    listen [::]:80;
    server_name autoconfig.myserver.tld;
    root /srv/automx/instance;

    access_log /var/log/nginx/autoconfig.myserver.tld-access.log;
    error_log /var/log/nginx/autoconfig.myserver.tld-error.log;

    location /mail/config-v1.1.xml {
        include uwsgi_params;
        uwsgi_pass automx;
    }
}
```
You can see that the autoconfig utility is listening on Port 80, which will no longer work if you completely block access to this port.

**Reply by CyberfoxDK** [reply]
I would not do so on the webserver itself, but on a dedicated firewall.
The most important thing for me, is that i will be able to renew the certificate even if it expires. I did not believe that was a possibility earlier, so it makes things much more simple.

If i were to do it on the server itself, i would probably create some iptable or netfilter firewall rule to block access to port 80, instead of messing with the code config (i am not a programmer) which i have limited knowledge about.
This could also be done as part of the cronjob, and disable the firewall rule blocking port 80, before the renewel attempt was executed.
Once the renewal was done of failed, the firewall rule could be activated again.

**Reply by nicosqc** [reply]
If you do not allow access of Port 80 through the firewall you will have issues using the automatic mail configuration features most mail clients provide, so your users will have to manually type in the IMAP- and SMTP-server details, as well as knowing the port for your mail server. A mail client will look in the preconfigured destinations using HTTP and Port 80 and since this is on the client side this will inevitably break.

[Autoconfiguration Utility automx2](https://github.com/rseichter/automx2)

You can see for yourself when using Outlook on Windows; open Outlook and then in the task bar right click on the Outlook symbol while holding down the CTRL-key. Select [Test E-mail AutoConfiguration](https://answers.microsoft.com/en-us/msoffice/forum/all/mailboxes-test-e-mail-autoconfiguration/19f9c90a-4640-46c4-a574-dbec29bdb8ba) with and without the firewall rule in place and check if it makes a difference. Maybe I am wrong and it works regardless, the Nginx configuration does suggest otherwise.

**Reply by CyberfoxDK** [reply]
You are absolutely correct, port 80 is needed for autocorrect. 
(Worked as an Exchange architech some years ago)
But in my case, this is blocked on the perimeter firewall.
Your point though, makes my suggestion to use iptable on the host to deny access to port 80 obsolete...  if you need autoconfig, you would need port 80 to be accessible.

**Reply by nicosqc** [reply]
If you have to manually open port 80 in an external firewall to do a renew of the certificate but you are able to automatically add or remove DNS records instead, you could have a look at the certbot docs for the ``--manual`` option [here](https://eff-certbot.readthedocs.io/en/stable/using.html#manual) and use the ``--preferred-challenges dns`` in a set of scripts or tweak the firewall to allow ``https://acme-v02.api.letsencrypt.org/directory`` as this is the [default server](https://eff-certbot.readthedocs.io/en/stable/using.html#changing-the-acme-server) that challenges are performed against (which you have already done from my understanding).
